### PR TITLE
Clean up and adapt to new code

### LINKED ISSUE #116: arizona_tpl_compile.erl: improve on try-catch
Was
```
        {% FIXME: This implementation wont work, because it     }
        {%        needs to be compiled in tokens.               }
        {%        The 'merl:tsubst' should be used to fix this. }
```

in `counter/1`.

### LINKED ISSUE #117: arizona_tpl_scan.erl: fix post-fun jump
Was
```
    % FIXME: If closing_tag we should go to 'scan'
    %        instead of 'scan_attr_key'.
```

in `scan_tag_name/3`.

### LINKED ISSUE #118: Remove as many `_to_atom` as possible
Replacing them with `_to_existing_atom`, where applicable.

### LINKED ISSUE #123: arizona_tpl_scan/parse.erl: optimizations
_1 (parse)
```
% TODO: Do not reverse.
```

in `do_parse_tag/3`

_2 (scan)
```
    % TODO: Location.
```

in `string/1`

### LINKED ISSUE #148: Exclude arizona_tpl_scan, arizona_tpl_parse, and arizona_tpl_compile
Depends on:
- https://github.com/arizona-framework/arizona/issues/142
- https://github.com/arizona-framework/arizona/issues/144
- https://github.com/arizona-framework/arizona/issues/145

### PR DESCRIPTION
**Pull request by williamthome**
# Description

This PR removes "old" modules:
- arizona_tpl_scan
- arizona_tpl_parse
- arizona_tpl_compile
- arizona_tpl_render

In favor of these new ones:
- arizona_template_scanner
- arizona_template_parser
- arizona_template_compiler
- arizona_template_renderer

It also adapts Erlang and Javascript files to the new code.

It's working as expected in the `arizona_example` project:
arizona_render_via_new_modules

Closes #116. Closes #117. Closes #118. Closes #123. Closes #148.

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/arizona-framework/arizona/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -127,15 +127,15 @@ handle_event(<<"decr">>, #{}, #{assigns := Assigns} = Socket) ->
 %% --------------------------------------------------------------------
 
 counter(Macros) ->
-    ?ARIZONA_LIVEVIEW(~s"""
+    ?ARIZONA_LIVEVIEW(~"""
```
Comment (williamthome) [inline README.md:130]: I was not aware that `~s"""` returns an `iolist()` and `~"""` a UTF-8 binary. For me, both would return a UTF-8 binary. This change is in README.md, which needs a complete revisit. Anyway, I'm changing here to remember this since the scanner expects a binary to be parsed.

File: README.md [hunk 2]
```diff
@@ -127,15 +127,15 @@ handle_event(<<"decr">>, #{}, #{assigns := Assigns} = Socket) ->
...
     <div :stateful>
         <div>Count: {_@count}</div>
         <.button event={_@event} text={_@btn_text} />
     </div>
     """).
 
 button(Macros) ->
-    ?ARIZONA_LIVEVIEW(~s"""
+    ?ARIZONA_LIVEVIEW(~"""
```
Comment (williamthome) [inline README.md:138]: Same here about `~s"""`  vs `~"""`.

File: priv/static/assets/js/arizona-worker.js [hunk 3]
```diff
@@ -111,32 +111,39 @@ function handleEvent(data) {
...
+function applyPatch([[, ...target], changes]) {
+  const tree = getTargetTree(target, state.tree[1]);
   changes.forEach((c) => {
     applyChanges(c, tree);
   });
-  const html = tree.flat(Infinity).join("");
-  return { target, html };
+  const html = zip(tree[0], tree[1]);
+  return { target: [0, ...target], html };
+}
```
Comment (williamthome) [inline priv/static/assets/js/arizona-worker.js:121]: A brief note here...
Now, block IDs always start with zero, and zero is the "root"/main block.
The server needs this information to handle sockets and renders, but the client "patch" doesn't.
What happens here is: I'm suppressing the zero in the function args and prefixing it to the return of the function in the 'target' prop to get the correct element in the DOM.
Currently, I don't have a solution that works on both sides without this "workaround" on the client side.
Anyway, it works :smile:

File: priv/static/assets/js/arizona-worker.js [hunk 4]
```diff
@@ -111,32 +111,39 @@ function handleEvent(data) {
...
+
+function getTargetTree(path, tree) {
+  if (path.length === 1) {
+    return tree[path[0]];
+  } else {
+    const [i, ...rest] = path;
+    return getTargetTree(rest, tree[i]);
+  }
 }
```
Comment (williamthome) [inline priv/static/assets/js/arizona-worker.js:130]: There are no changes here, just functions reorder.

File: priv/static/assets/js/arizona-worker.js [hunk 5]
```diff
@@ -111,32 +111,39 @@ function handleEvent(data) {
...
-    return tree[path[0]];
-  } else {
-    const [i, ...rest] = path;
-    return getTargetTree(rest, tree[i]);
+function zip(staticArr, dynamicArr) {
+  let str = "";
+  for (let i = 0; i < Math.max(staticArr.length, dynamicArr.length); i++) {
+    str += `${staticArr[i] || ""}${dynamicArr[i] || ""}`;
   }
+  return str;
```
Comment (williamthome) [inline priv/static/assets/js/arizona-worker.js:146]: The same thing happens now on the server side: We "zip" static and dynamic lists.
`tree[0]` is a static list and `tree[1]` is dynamic.
I was unaware that `static` is a reserved word in Javascript, so I renamed the function args `staticArr` and `dynamicArr` instead of just `static` and `dynamic`.
> Arr = array.

File: priv/static/assets/js/arizona.js [hunk 6]
```diff
@@ -107,10 +106,12 @@ globalThis["arizona"] = (() => {
...
     if (!opts.target && this instanceof HTMLElement) {
-      opts.target = this.getAttribute("arz-target");
+      opts.target = this.getAttribute("arizona-target");
     }
+    const target = document.querySelector(opts.target);
+    const arizonaId = target?.getAttribute("arizona-id") || "[0]";
     callback && subscribeOnce(event, callback, opts);
-    worker.postMessage({ target: opts.target || "root", event, payload });
+    worker.postMessage({ target: arizonaId, event, payload });
   }
```
Comment (williamthome) [inline priv/static/assets/js/arizona.js:115]: `arz-id` was changed to `arizona-id` and `arz-target` to `arizona-target`.

File: src/arizona_socket.erl [hunk 7]
```diff
@@ -52,7 +52,7 @@ new(View, Assigns) ->
         view => View,
         assigns => Assigns,
         events => [],
-        changes => #{}
+        changes => ordsets:new()
```
Comment (williamthome) [inline src/arizona_socket.erl:55]: `changes` is now a list of unique atoms.

File: src/arizona_template_compiler.erl [hunk 8]
```diff
@@ -125,22 +125,23 @@ compile_expr(ExprStr, Loc, T, State) ->
...
+    Macros = State#state.macros,
+    MacrosEnv = [{K, merl:term(V)} || K := V <- Macros],
     AttrsEnv = lists:map(fun
         ({K, {expr, _KLoc, KExpr}}) ->
-            {binary_to_existing_atom(K, utf8), merl:quote(KExpr)};
+            KForm = merl:tsubst(merl:quote(KExpr), MacrosEnv),
+            {binary_to_existing_atom(K, utf8), KForm};
         ({K, {text, _KLoc, KTxt}}) ->
-            {binary_to_existing_atom(K, utf8),
+            {binary_to_atom(K, utf8),
```
Comment (williamthome) [inline src/arizona_template_compiler.erl:135]: It's not possible to use `binary_to_existing_atom` in text attributes.

File: README.md (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply README.md:130]: I wouldn't worry too much about the README now. I'm sure some other stuff's broken already. And the less changes we make in pull requests the easier they are to review 😄

File: priv/static/assets/js/arizona-worker.js
Comment (paulo-ferraz-oliveira) [inline priv/static/assets/js/arizona-worker.js:1]: I'm not reviewing this file too much, since I'm touch the JavaScript layer locally and it's possible it gets turned around. In any case, I'll have to look at it closer, soon, so will leave my remarks for a later moment. Feel free to resolve this conversation upon reading it.

File: priv/static/assets/js/arizona.js (hunk 6 above)
Comment (paulo-ferraz-oliveira) [reply priv/static/assets/js/arizona.js:115]: We had an issue for such considerations (https://github.com/arizona-framework/arizona/issues/115). I'd prefer it if we did smaller incremental changes, since they'd be easier to follow: in any case, since this is now done, feel free to cram that issue into this pull request too.

File: src/arizona_html.erl [hunk 9]
```diff
@@ -31,7 +30,3 @@ to_safe(V) when is_integer(V) ->
     integer_to_binary(V, 10);
 to_safe(V) when is_float(V) ->
     io_lib:format("~p", [V]).
-
--spec safe_types() -> [binary | atom | integer | float].
-safe_types() ->
-    [binary, atom, integer, float].
```
Comment (paulo-ferraz-oliveira) [inline src/arizona_html.erl:37]: Where did this go?

File: src/arizona_live_view.erl [hunk 10]
```diff
@@ -36,9 +36,14 @@ Live view.
...
-    when Macros :: arizona_template_compiler:macros(),
-         Tree :: arizona_tpl_parse:tree().
+-callback render(Macros1) -> Result
+    when Macros1 :: arizona_template_compiler:macros(),
+         Result :: {ok, {Parsed, Macros2}}
+                 | {error, ErrReason},
+         Parsed :: [arizona_template_parser:element()],
+         Macros2 :: arizona_template_compiler:macros(),
+         ErrReason :: arizona_template_scanner:error_reason()
+                    | arizona_template_parser:error_reason().
```
Comment (paulo-ferraz-oliveira) [inline src/arizona_live_view.erl:46]: Does this mean we'd not know if the error comes from one of the other, and force consumer to check? I think `{error, {scanner, Error}} | {error, {parser, Error}}` would be helpful (?)

File: src/arizona_live_view.erl [hunk 11]
```diff
@@ -69,12 +74,24 @@ put_macro(Key, Value, Macros) ->
...
+-spec parse_str(Str, Macros1) -> Result
     when Str :: string() | binary(),
-         Macros :: arizona_template_compiler:macros().
+         Macros1 :: arizona_template_compiler:macros(),
+         Result :: {ok, {Parsed, Macros2}}
+                 | {error, ErrReason},
+         Parsed :: [arizona_template_parser:element()],
+         Macros2 :: arizona_template_compiler:macros(),
+         ErrReason :: arizona_template_scanner:error_reason()
+                    | arizona_template_parser:error_reason().
```
Comment (paulo-ferraz-oliveira) [inline src/arizona_live_view.erl:85]: Same consideration as above.

File: src/arizona_live_view.erl [hunk 12]
```diff
@@ -87,10 +104,15 @@ parse_str(Str, Macros) ->
...
     when Mod :: module(),
-         Macros :: arizona_template_compiler:macros(),
-         Tree :: arizona_tpl_parse:tree().
+         Macros1 :: arizona_template_compiler:macros(),
+         Result :: {ok, {Parsed, Macros2}}
+                 | {error, ErrReason},
+         Parsed :: [arizona_template_parser:element()],
+         Macros2 :: arizona_template_compiler:macros(),
+         ErrReason :: arizona_template_scanner:error_reason()
+                    | arizona_template_parser:error_reason().
```
Comment (paulo-ferraz-oliveira) [inline src/arizona_live_view.erl:115]: Same consideration as above.

File: src/arizona_socket.erl [hunk 13]
```diff
@@ -124,77 +124,5 @@ push_event(Name, Payload, #{events := Events} = Socket) ->
...
--ifdef(TEST).
--compile([export_all, nowarn_export_all]).
--include_lib("eunit/include/eunit.hrl").
-
-render_block_test() ->
-    ?assertEqual(
-       rendered(), arizona_tpl_render:render_block(block(#{}), #{
-            title => <<"Arizona">>,
-            view_count => 0,
-            decr_btn_text => <<"Decrement">>})).
```
Comment (paulo-ferraz-oliveira) [inline src/arizona_socket.erl:143]: Unless we're doing this for good reason, I'd not start moving EUnit to CT at this moment (especially given this PR's size, already).

Comment (paulo-ferraz-oliveira) [reply src/arizona_socket.erl:143]: (remember we have https://github.com/arizona-framework/arizona/issues/155)

File: src/arizona_template_compiler.erl [hunk 14]
```diff
@@ -125,22 +125,23 @@ compile_expr(ExprStr, Loc, T, State) ->
...
     end, State#state.attributes),
     AttrsTree = merl:tsubst(ExprTree, AttrsEnv),
-    Macros = State#state.macros,
-    MacrosEnv = [{K, merl:term(V)} || K := V <- Macros],
     MacrosTree = merl:tsubst(AttrsTree, MacrosEnv),
-    AllVars = merl:template_vars(merl:template(MacrosTree)),
-    case AllVars -- maps:keys(Macros) of
-        [] when Eval ->
+    Vars = merl:template_vars(merl:template(MacrosTree)),
+    case Vars =:= [] andalso Eval of
```
Comment (paulo-ferraz-oliveira) [inline src/arizona_template_compiler.erl:141]: Did you try `case {Vars, Eval} of` and then a different set of case clauses, for readability?

File: src/arizona_template_compiler.erl [hunk 15]
```diff
@@ -330,7 +336,7 @@ push_js_scripts([]) ->
 js_scripts() ->
     Scripts = case arizona_cfg:endpoint() of
         #{live_reload := true} ->
-            [live_reload_script() | required_js_script()];
+            required_js_script() ++ [live_reload_script()];
```
Comment (paulo-ferraz-oliveira) [inline src/arizona_template_compiler.erl:339]: Is the order relevant, here? Or why the change?

**Review by paulo-ferraz-oliveira — COMMENTED** [review]
Changes seem sane. Feel free to resolve conversations I looked at, with 👍 or similar reactions. A few questions still remain, though, the most important comment being "Let's try to have smaller pull requests, focused on a given change".

File: src/arizona_template_compiler.erl (hunk 15 above)
Comment (williamthome) [reply src/arizona_template_compiler.erl:339]: Yes, it's relevant because `arizona-live-reload.js` depends on `arizona.js` (which is included in `required_js_script()`).

File: src/arizona_template_compiler.erl (hunk 14 above)
Comment (williamthome) [reply src/arizona_template_compiler.erl:141]: In that case, I need a boolean value, so I don't see a readability issue here.

File: src/arizona_socket.erl (hunk 13 above)
Comment (williamthome) [reply src/arizona_socket.erl:143]: This test is duplicated from the new renderer module, so it's irrelevant.

File: src/arizona_live_view.erl (hunk 10 above)
Comment (williamthome) [reply src/arizona_live_view.erl:46]: Sounds good. I'll change it to `{error, {scanner, Error}} | {error, {parser, Error}}`.

File: src/arizona_template_compiler.erl (hunk 14 above)
Comment (paulo-ferraz-oliveira) [reply src/arizona_template_compiler.erl:141]: Check

```erlang
    case Vars =:= [] andalso Eval of
        true ->
            ... something;
        false ->
            ... something else
    end
```

vs.

```erlang
    case {Vars, Eval} of
        {[], true} ->
            ... something;
        _Other ->
            ... something else
    end
```

In the first bit of code I have to reason on the shortcut. In the second bit of code I know exactly (and in the same line) what `Vars` and `Eval` and both expected to be.

Better still would be to name that condition (`Vars =:= [] andalso Eval`) because with a bit less context you won't know what it stands for.

File: src/arizona_html.erl (hunk 9 above)
Comment (williamthome) [reply src/arizona_html.erl:37]: It is no longer necessary as it was used in the old compiler.

File: src/arizona_template_compiler.erl (hunk 15 above)
Comment (paulo-ferraz-oliveira) [reply src/arizona_template_compiler.erl:339]: Right. Did we think about using `defer` or `async` in that case, to minimize dependency-related issues?

Comment (williamthome) [reply src/arizona_template_compiler.erl:339]: Maybe the `morphdom` lib could be loaded `async`, but I'm unsure because If the consumer needs to update the DOM immediately, this can cause an issue. I'm against using `defer` to load the `arizona.js` since we want the WebSocket connected as fast as possible. I'm for checking this type of feature when benchmarks and tests are implemented in JS files.

File: src/arizona_template_compiler.erl (hunk 14 above)
Comment (williamthome) [reply src/arizona_template_compiler.erl:141]: Thanks for the suggestion; I agree. The second code is much easier for my eyes to follow visually, while the first code requires a little more attention to understand.

File: priv/static/assets/js/arizona.js (hunk 6 above)
Comment (williamthome) [reply priv/static/assets/js/arizona.js:115]: I saw it, but it's not related to naming these attributes but to creating new ones to improve "patch" performance. It's a future stuff. I'll rename/update its title.

File: src/arizona_live_view.erl (hunk 10 above)
Comment (williamthome) [reply src/arizona_live_view.erl]: [37c2b46](https://github.com/arizona-framework/arizona/pull/177/commits/37c2b46e002b0f6f819bf25369706f0e3b436912)

File: src/arizona_template_compiler.erl (hunk 14 above)
Comment (williamthome) [reply src/arizona_template_compiler.erl]: [59d551a](https://github.com/arizona-framework/arizona/pull/177/commits/59d551adf070235c838d5651763efe90f5d2ae97)

**Review by paulo-ferraz-oliveira — APPROVED** [review]
